libs/contrib/Data/String/Parser.idr

```diff
@@ -0,0 +1,151 @@
+||| A simple parser combinator library for strings. Inspired by attoparsec zepto.
+module Data.String.Parser
+
+import Control.Monad.Identity
+import Control.Monad.Trans
+
+import Data.Strings
+
+%default partial
+
+||| The input state, pos is position in the string and maxPos is the length of the input string.
+public export
+record State where
+    constructor S
+    input : String
+    pos : Int
+    maxPos : Int
+
+Show State where
+    show s = "(" ++ show s.input ++", " ++ show s.pos ++ ", " ++ show s.maxPos ++")"
+
+||| A debug helper to print out the input state during parsing.
```

Comment:
Was this function left here intentionally? I only ask because this function breaks referential transparency and, in my opinion, such functions should be marked in a way that makes that obvious. For example by naming it `unsafeDumpState` or putting it in a module that include `Debug`, `Unsafe` (or similar) in its name.

There exist a function that could be used for this use-case, in `Debug.Trace`:
```
trace : (msg : String) -> (result : a) -> a
trace x val = unsafePerformIO (do putStrLn x; pure val)
```

There could probably also be added a function like the following (name is copied from Haskell's `Debug.Trace` module):
```
traceShowId : Show a => (result : a) -> a
traceShowId val = trace (show val) val
```

Depending on where you'd want to use `dumpState`, maybe it makes sense to have a function like:
```
dumpState : HasIO io => ParseT io ()
dumpState = P $ \s => do
  printLn s
  pure (OK () s)
```

Reply:
I'll remove that, it's not exported anyway.